scripts/gke/iam_patch.py

```diff
@@ -0,0 +1,128 @@
+
+
+def iam_policy_to_dict(bindings):
+  """
+  iam_policy_to_dict takes an iam policy binding in the GCP API format and converts it into a python dict
+  so that it can be easily updated
+  """
+  bindings_dict = dict()
+  for binding in bindings:
+    role = binding['role']
+    bindings_dict[role] = set()
+    for member in binding['members']:
+      bindings_dict[role].add(member)
+  return bindings_dict
+
+
+def iam_dict_to_policy(bindings_dict):
+  """
+  iam_dict_to_policy takes an iam policy binding in the dict format and converts it into GCP API format
+  so that it can be sent to GCP IAM API for an update
+  """
+  bindings = []
+  for k, v in bindings_dict.items():
+    bindings.append({"role": k, "members": list(v)})
+  return bindings
+
+def apply_iam_bindings_patch(current_policy, bindings_patch, action):
+  """
+  Patches the current policy with the supplied patch.
+  action can be add or remove.
+  """
+  for member, roles in bindings_patch['gcpIamPolicy']['bindings'].items():
+    for role in roles:
+      if role not in current_policy.keys():
+        current_policy[role] = set()
+      for member_email in bindings_patch['gcpIamPolicy']['members'][member]:
+        if action == "add":
+          current_policy[role].add(member_email)
+        else:
+          current_policy[role].discard(member_email)
+  return current_policy
+
+
+def patch_iam_policy(args):
+  """
+  Fetches the current IAM policy, patches it with the bindings supplied
+  in --iam_bindings_file and updates the new iam policy
+  """
+  current_policy = get_current_iam_policy(args.project)
+  logging.info("Current IAM Policy")
+  logging.info(yaml.dump(current_policy, default_flow_style=False, default_style=''))
+  current_policy_bindings_dict = iam_policy_to_dict(current_policy['bindings'])
+  with open(args.iam_bindings_file) as iam_bindings_file:
+    bindings_patch = yaml.load(iam_bindings_file.read())
+  current_policy_bindings_dict = apply_iam_bindings_patch(current_policy_bindings_dict, bindings_patch, args.action)
+  current_policy['bindings'] = iam_dict_to_policy(current_policy_bindings_dict)
+  logging.info("Updated Policy")
+  logging.info("\n" + yaml.dump(current_policy, default_flow_style=False, default_style=''))
+  updated_policy_file = tempfile.NamedTemporaryFile(delete=False)
+  with open(updated_policy_file.name, 'w') as f:
+    yaml.dump(current_policy, f, default_flow_style=False)
+  logging.debug("Temp file %s", updated_policy_file.name)
+  if not args.dry_run:
+    subprocess.call(["gcloud", "projects", "set-iam-policy", args.project, updated_policy_file.name])
+  else:
+    logging.info("Skipping patching the IAM policy because --dry_run was set")
+
+
+if __name__ == "__main__":
+  logging.getLogger().setLevel(logging.INFO)
+  args = parse_args()
+
+  if args.action not in ["add", "remove"]:
+    raise ValueError("invalid --action. Valid values are add, remove")
+
+  for i in range(5):
```

Comment:
Can we use the retrying module?

Reply:
I intentionally tried to avoid using `retrying` module because this script will be run on an end-users computer and I wanted to avoid an additional step of doing a `pip install`